File: frontend/src/utils/oddsTriggerLookup.ts
/**
 * Odds Trigger Lookup Utility
 * Calculates minimum acceptable betting odds based on strategy win rate
 */

interface OddsLookupEntry {
  win_rate_pct: number;
  win_probability: number;
  decimal_odds: number;
  american_odds_exact: number;
  american_odds_rounded: number;
  min_acceptable_odds: string;
}

// Pre-calculated lookup table (loaded from public/odds_trigger_lookup.json)
let oddsLookupTable: OddsLookupEntry[] = [];

/**
 * Load the odds lookup table from public folder
 */
export async function loadOddsLookupTable(): Promise<void> {
  try {
    const response = await fetch('/odds_trigger_lookup.json');
    oddsLookupTable = await response.json();
  } catch (error) {
    console.error('Failed to load odds trigger lookup table:', error);
    oddsLookupTable = [];
  }
}

/**
 * Calculate minimum acceptable odds for a given win rate
 * Uses interpolation for win rates between table entries
 *
 * @param winRate - Win rate as percentage (e.g., 53.4 for 53.4%)
 * @returns Minimum acceptable odds in American format (e.g., "-115", "+120")
 */
export function getMinAcceptableOdds(winRate: number): string {
  if (!oddsLookupTable || oddsLookupTable.length === 0) {
    // Fallback calculation if table not loaded
    return calculateOddsDirectly(winRate);
  }

  // Handle edge cases
  if (winRate < 50) return '+100'; // Below 50% = underdog territory
  if (winRate >= 75) return '-300'; // 75%+ = max in our table

  // Find the two closest entries in the lookup table
  let lowerEntry: OddsLookupEntry | null = null;
  let upperEntry: OddsLookupEntry | null = null;

  for (let i = 0; i < oddsLookupTable.length - 1; i++) {
    if (winRate >= oddsLookupTable[i].win_rate_pct && winRate <= oddsLookupTable[i + 1].win_rate_pct) {
      lowerEntry = oddsLookupTable[i];
      upperEntry = oddsLookupTable[i + 1];
      break;
    }
  }

  // Exact match found
  if (lowerEntry && lowerEntry.win_rate_pct === winRate) {
    return lowerEntry.min_acceptable_odds;
  }

  // Interpolate between two entries
  if (lowerEntry && upperEntry) {
    const ratio = (winRate - lowerEntry.win_rate_pct) / (upperEntry.win_rate_pct - lowerEntry.win_rate_pct);
    const interpolatedOdds = lowerEntry.american_odds_rounded +
                             ratio * (upperEntry.american_odds_rounded - lowerEntry.american_odds_rounded);

    // Round to nearest 5 (standard sportsbook practice)
    const rounded = Math.round(interpolatedOdds / 5) * 5;
    return rounded >= 0 ? `+${rounded}` : `${rounded}`;
  }

  // Fallback to direct calculation
  return calculateOddsDirectly(winRate);
}

/**
 * Direct calculation of minimum acceptable odds (fallback)
 *
 * @param winRate - Win rate as percentage
 * @returns Minimum acceptable odds in American format
 */
function calculateOddsDirectly(winRate: number): string {
  const winProbability = winRate / 100;
  const decimalOdds = 1 / winProbability;

  let americanOdds: number;

  if (winProbability > 0.5) {
    // Favorite (negative odds)
    americanOdds = -100 / (decimalOdds - 1);
  } else if (winProbability < 0.5) {
    // Underdog (positive odds)
    americanOdds = 100 * (decimalOdds - 1);
  } else {
    // Even money
    americanOdds = 100;
  }

  // Round to nearest 5
  const rounded = Math.round(americanOdds / 5) * 5;
  return rounded >= 0 ? `+${rounded}` : `${rounded}`;
}

/**
 * Get odds with color coding for UI display
 *
 * @param winRate - Win rate as percentage
 * @returns Object with odds and color class
 */
export function getOddsWithColor(winRate: number): { odds: string; colorClass: string } {
  const odds = getMinAcceptableOdds(winRate);

  // Determine color based on odds value
  let colorClass = 'text-white';

  if (odds.startsWith('+')) {
    // Positive odds (underdog) - green
    colorClass = 'text-green-400';
  } else if (odds.startsWith('-')) {
    const value = Math.abs(parseInt(odds));
    if (value >= 200) {
      // Heavy favorite - red
      colorClass = 'text-red-400';
    } else if (value >= 150) {
      // Strong favorite - orange
      colorClass = 'text-orange-400';
    } else {
      // Slight favorite - yellow
      colorClass = 'text-yellow-400';
    }
  }

  return { odds, colorClass };
}

/**
 * Format odds for display with proper styling
 *
 * @param winRate - Win rate as percentage
 * @returns Formatted odds string with sign
 */
export function formatOddsDisplay(winRate: number | undefined): string {
  if (!winRate || winRate < 50) {
    return 'N/A';
  }

  return getMinAcceptableOdds(winRate);
}

File: backend/odds_trigger_lookup.py
"""
Odds Trigger Lookup Calculator
Calculates minimum acceptable betting odds based on strategy win rate

Formula:
- If win_rate > 50%: American Odds = -100 / (decimal - 1) where decimal = 1 / win_probability
- If win_rate < 50%: American Odds = 100 * (decimal - 1) where decimal = 1 / win_probability
- If win_rate = 50%: American Odds = +100 (even money)
"""

import json
import csv
from pathlib import Path


def calculate_break_even_odds(win_rate_pct: float) -> dict:
    """
    Calculate minimum acceptable odds for a given win rate

    Args:
        win_rate_pct: Win rate as percentage (e.g., 53.4 for 53.4%)

    Returns:
        Dictionary with american_odds, decimal_odds, win_rate
    """
    win_probability = win_rate_pct / 100

    # Calculate fair decimal odds
    decimal_odds = 1 / win_probability

    # Calculate American odds
    if win_probability > 0.50:
        # Favorite (negative odds)
        american_odds = -100 / (decimal_odds - 1)
    elif win_probability < 0.50:
        # Underdog (positive odds)
        american_odds = 100 * (decimal_odds - 1)
    else:
        # Even money
        american_odds = 100

    # Round to nearest 5 (standard sportsbook practice)
    american_odds_rounded = round(american_odds / 5) * 5

    return {
        "win_rate_pct": round(win_rate_pct, 1),
        "win_probability": round(win_probability, 4),
        "decimal_odds": round(decimal_odds, 2),
        "american_odds_exact": round(american_odds, 1),
        "american_odds_rounded": int(american_odds_rounded),
        "min_acceptable_odds": f"{'+' if american_odds_rounded >= 0 else ''}{int(american_odds_rounded)}"
    }


def generate_lookup_table():
    """Generate complete lookup table from 50% to 75% win rate"""

    lookup_data = []

    # Generate for every 0.1% from 50% to 75%
    win_rate = 50.0
    while win_rate <= 75.0:
        odds_data = calculate_break_even_odds(win_rate)
        lookup_data.append(odds_data)
        win_rate += 0.1

    return lookup_data


def save_to_json(data: list, filepath: str):
    """Save lookup table as JSON"""
    with open(filepath, 'w') as f:
        json.dump(data, f, indent=2)
    print(f"[OK] Saved JSON: {filepath}")


def save_to_csv(data: list, filepath: str):
    """Save lookup table as CSV"""
    fieldnames = data[0].keys()
    with open(filepath, 'w', newline='') as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(data)
    print(f"[OK] Saved CSV: {filepath}")


def create_key_milestones_table():
    """Create a simplified table with key win rate milestones"""
    key_rates = [
        50.0, 51.0, 52.0, 52.5, 53.0, 53.4, 54.0, 55.0,
        56.0, 57.0, 58.0, 59.0, 60.0, 61.0, 62.0, 63.0,
        64.0, 65.0, 66.0, 67.0, 68.0, 69.0, 70.0, 75.0
    ]

    return [calculate_break_even_odds(rate) for rate in key_rates]


def main():
    """Generate all lookup files"""
    print("=" * 80)
    print("ODDS TRIGGER LOOKUP TABLE GENERATOR")
    print("=" * 80)
    print("\nCalculating break-even odds for win rates 50.0% to 75.0%...\n")

    # Generate complete lookup table (every 0.1%)
    full_data = generate_lookup_table()

    # Generate key milestones table
    key_data = create_key_milestones_table()

    # Save complete data
    save_to_json(full_data, "data/odds_trigger_full_lookup.json")
    save_to_csv(full_data, "data/odds_trigger_full_lookup.csv")

    # Save key milestones
    save_to_json(key_data, "data/odds_trigger_key_milestones.json")
    save_to_csv(key_data, "data/odds_trigger_key_milestones.csv")

    print(f"\n[OK] Generated {len(full_data)} complete entries")
    print(f"[OK] Generated {len(key_data)} key milestone entries")

    # Print sample output
    print("\n" + "=" * 80)
    print("SAMPLE OUTPUT (Key Milestones):")
    print("=" * 80)
    print(f"\n{'Win Rate':<12} {'Probability':<14} {'Decimal':<10} {'American':<12} {'Min Accept'}")
    print("-" * 80)

    for entry in key_data[:15]:  # Show first 15
        print(f"{entry['win_rate_pct']:<11}% {entry['win_probability']:<14} {entry['decimal_odds']:<10} "
              f"{entry['american_odds_exact']:<12.1f} {entry['min_acceptable_odds']}")

    print("\n" + "=" * 80)
    print("FILES CREATED:")
    print("=" * 80)
    print("1. data/odds_trigger_full_lookup.json    - Complete table (every 0.1%)")
    print("2. data/odds_trigger_full_lookup.csv     - Complete table (CSV format)")
    print("3. data/odds_trigger_key_milestones.json - Key win rates only")
    print("4. data/odds_trigger_key_milestones.csv  - Key win rates (CSV format)")
    print("=" * 80)


if __name__ == "__main__":
    main()
